### PR TITLE
Implementing BPMN editor using d3

### PR DESCRIPTION
**Pull request by NatashaWso2**

### CODE REVIEW

File: modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java [hunk 1]
```diff
@@ -1418,6 +1422,70 @@ public String getFlowchart(String flowchartPath) {
...
+
+                String newProcessContent = xmlToString(processXMLContent);
+                processAsset.setContent(newProcessContent);
+                reg.put(processPath, processAsset);
+
+                Resource storedProcessAsset = reg.get(processPath);
+                processId = storedProcessAsset.getUUID();
+            }
+        } catch (Exception e) {
+            log.error(e.getMessage());
```
Comment (nandika) [inline modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java]: Need to log the exception properly here in the format. log.error( "Error Info " , e)

File: modules/extensions/process/publisher/process/themes/default/js/bpmnEditor/activityEventHandler.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+             bpmnProcessDiagram.links.push(newArrow);
+             bpmnProcessDiagram.drawGraph();
+         }
+     } else {
+            // In the same node
+            if (stateOfProcessDiagram.dragActive) {               
+                stateOfProcessDiagram.dragActive = false;
+            } else {
+               
+                if (d3.event.shiftKey) {
```
Comment (nandika) [inline modules/extensions/process/publisher/process/themes/default/js/bpmnEditor/activityEventHandler.js]: move logic to if condition

File: modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java (hunk 1 above)
Comment (NatashaWso2) [inline modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java]: Fixed with commit ab8cc47860ca05cc98ee427419202b963c3dfe6b

File: modules/extensions/process/publisher/process/themes/default/js/bpmnEditor/activityEventHandler.js (hunk 2 above)
Comment (NatashaWso2) [inline modules/extensions/process/publisher/process/themes/default/js/bpmnEditor/activityEventHandler.js]: Fixed with commit ab8cc47

File: modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java (hunk 1 above)
Comment (NatashaWso2) [inline modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java]: Fixed with commit ab8cc47860ca05cc98ee427419202b963c3dfe6b

File: modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java [hunk 3]
```diff
@@ -1418,6 +1422,70 @@ public String getFlowchart(String flowchartPath) {
         return flowchartString;
     }
 
+    /**
+     *
+     * @param processName
```
Comment (pamod) [inline modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java]: It would be good to have params defined

Comment (NatashaWso2) [inline modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java]: Params are defined and fixed with commit 762e1814389e8920882601d4b8d7d2363667fe97

File: modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java [hunk 4]
```diff
@@ -1418,6 +1422,70 @@ public String getFlowchart(String flowchartPath) {
...
+     * @param processVersion
+     * @param bpmndesignJson
+     * @return the bpmn diagram json string for a process
+     */
+    public String uploadBpmnDesign(String processName, String processVersion, String bpmndesignJson) {
+        String processId = "NA";
+        log.debug("Creating bpmn design...");
```
Comment (lalaji) [inline modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java]: add log debug entries checking if debug level logs enabled.

File: modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java [hunk 5]
```diff
@@ -1418,6 +1422,70 @@ public String getFlowchart(String flowchartPath) {
...
+                processXMLContent.getElementsByTagName("bpmnDesign").item(0).getFirstChild().setTextContent(bpmnDesignContentPath);
+
+                String newProcessContent = xmlToString(processXMLContent);
+                processAsset.setContent(newProcessContent);
+                reg.put(processPath, processAsset);
+
+                Resource storedProcessAsset = reg.get(processPath);
+                processId = storedProcessAsset.getUUID();
+            }
+        } catch (Exception e) {
```
Comment (lalaji) [inline modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java]: It's recommended to catch the typed exceptions,rather catch as a general exception.And in here,we only log the exception.Is it sufficient or aren't we need to re throw it?

File: modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java (hunk 4 above)
Comment (NatashaWso2) [inline modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java]: Fixed with commit 8f9e509c152e816bc0f6963f69ca4c95cf582562

File: modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java [hunk 6]
```diff
@@ -1418,6 +1422,70 @@ public String getFlowchart(String flowchartPath) {
...
+        try {
+            RegistryService registryService = ProcessCenterServerHolder.getInstance().getRegistryService();
+            if (registryService != null) {
+                UserRegistry reg = registryService.getGovernanceSystemRegistry();
+                bpmnDiagramPath = bpmnDiagramPath.substring("/_system/governance/".length());
+                Resource bpmnDiagramAsset = reg.get(bpmnDiagramPath);
+                bpmnDiagramString = new String((byte[]) bpmnDiagramAsset.getContent());
+            }
+        } catch (Exception e) {
+            log.error(e.getMessage());
```
Comment (mohanvive) [inline modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java]: Please log complete error trace, not only the error message.. BTW, any reason for catching generic exception ?

File: modules/extensions/process/publisher/process/apis/upload_bpmnEditorDiagram.jag [hunk 7]
```diff
@@ -0,0 +1,30 @@
+<%
+/*
+ *  Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (mohanvive) [inline modules/extensions/process/publisher/process/apis/upload_bpmnEditorDiagram.jag]: As per the commit, it seems above file is created in 2016. Then any reason for specifying as 2015 here ??

File: modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java (hunk 6 above)
Comment (NatashaWso2) [inline modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java]: I have fixed the above issue with commit b17269ac9ddbadd992ed11cfb796a9ad7e2af3fc. The specific exceptions are catched without catching the generic exception.

File: modules/extensions/process/publisher/process/apis/upload_bpmnEditorDiagram.jag (hunk 7 above)
Comment (NatashaWso2) [inline modules/extensions/process/publisher/process/apis/upload_bpmnEditorDiagram.jag]: Fixed with commit b17269ac9ddbadd992ed11cfb796a9ad7e2af3fc

File: modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java (hunk 5 above)
Comment (NatashaWso2) [inline modules/components/core/org.wso2.carbon.pc.core/src/main/java/org/wso2/carbon/pc/core/ProcessStore.java]: I have fixed the above issue with commit b17269a. The specific exceptions are catched without catching the generic exception.

**Pull request merged**